pkg/ccl/streamingccl/streamproducer/replication_stream_test.go

```diff
@@ -0,0 +1,311 @@
+		cancel: cancel,
+	}
+}
+
+// newReplicationHelper starts test server and configures it to have
+// active tenant.
+func newReplicationHelper(t *testing.T) (*replicationHelper, func()) {
+	ctx := context.Background()
+
+	// Start server
+	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{})
+
+	// Set required cluster settings.
+	_, err := db.Exec(`
+SET CLUSTER SETTING kv.rangefeed.enabled = true;
+SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1s';
+SET CLUSTER SETTING changefeed.experimental_poll_interval = '10ms'
+`)
+	require.NoError(t, err)
+
+	// Make changefeeds run faster.
+	resetFreq := changefeedbase.TestingSetDefaultFlushFrequency(50 * time.Microsecond)
+
+	// Start tenant server
+	tenantID := roachpb.MakeTenantID(10)
+	_, tenantConn := serverutils.StartTenant(t, s, base.TestTenantArgs{TenantID: tenantID})
+
+	// Sink to read data from.
+	sink, cleanupSink := sqlutils.PGUrl(t, s.ServingSQLAddr(), t.Name(), url.User(security.RootUser))
+
+	h := &replicationHelper{
+		sysServer: s,
+		sysDB:     sqlutils.MakeSQLRunner(db),
+		sink:      sink,
+		tenant: tenantState{
+			id:    tenantID,
+			codec: keys.MakeSQLCodec(tenantID),
+			sql:   sqlutils.MakeSQLRunner(tenantConn),
+		},
+	}
+
+	return h, func() {
+		resetFreq()
+		cleanupSink()
+		require.NoError(t, tenantConn.Close())
+		s.Stopper().Stop(ctx)
+	}
+}
+
+func TestReplicationStreamTenant(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	h, cleanup := newReplicationHelper(t)
+	defer cleanup()
+
+	h.tenant.sql.Exec(t, `
+CREATE DATABASE d;
+CREATE TABLE d.t1(i int primary key, a string, b string);
+CREATE TABLE d.t2(i int primary key);
+INSERT INTO d.t1 (i) VALUES (42);
+INSERT INTO d.t2 VALUES (2);
+`)
+
+	streamTenantQuery := fmt.Sprintf(
+		`CREATE REPLICATION STREAM FOR TENANT %d`, h.tenant.id.ToUint64())
+
+	t.Run("cannot-stream-tenant-from-tenant", func(t *testing.T) {
+		// Cannot replicate stream from inside the tenant
+		_, err := h.tenant.sql.DB.ExecContext(context.Background(), streamTenantQuery)
+		require.Error(t, err)
+	})
+
+	feed := h.StartReplication(t, streamTenantQuery)
+	defer feed.Close()
+
+	descr := catalogkv.TestingGetTableDescriptor(h.sysServer.DB(), h.tenant.codec, "d", "t1")
+	expected := encodeKV(t, h.tenant.codec, descr, 42)
+	observed := feed.ObserveKey(expected.Key)
+
+	require.True(t, expected.Value.EqualTagAndData(observed.kv.Value))
+
+	// Periodically, resolved timestamps should be published.
+	// Observe resolved timestamp that's higher than the previous value timestamp.
+	feed.ObserveResolved(observed.kv.Value.Timestamp)
+
+	// Update our row.
+	h.tenant.sql.Exec(t, `UPDATE d.t1 SET b = 'blah' WHERE i = 42`)
+	expected = encodeKV(t, h.tenant.codec, descr, 42, nil, "blah")
+
+	// Observe its changes.
+	updated := feed.ObserveKey(expected.Key)
+	require.True(t, expected.Value.EqualTagAndData(updated.kv.Value))
+	require.True(t, observed.kv.Value.Timestamp.Less(updated.kv.Value.Timestamp))
```

Comment:
nit: can we rename `observed` and `updated` to something like `firstObserved` and `secondObserved`? I think it would make reading this line a bit clearer.